astropy/utils/data_info.py

```diff
@@ -184,7 +184,82 @@ def _get_data_attribute(dat, attr=None):
     return str(val)
 
 
-class DataInfo:
+class InfoAttribute:
+    def __init__(self, attr, default=None):
+        self.attr = attr
+        self.default = default
+
+    def __get__(self, instance, owner_cls):
+        if instance is None:
+            return self
+
+        return instance._attrs.get(self.attr, self.default)
+
+    def __set__(self, instance, value):
+        if instance is None:
+            # This is an unbound descriptor on the class
+            raise ValueError('cannot set unbound descriptor')
+
+        instance._attrs[self.attr] = value
+
+
+class ParentAttribute:
+    def __init__(self, attr):
+        self.attr = attr
+
+    def __get__(self, instance, owner_cls):
+        if instance is None:
+            return self
+
+        return getattr(instance._parent, self.attr)
+
+    def __set__(self, instance, value):
+        if instance is None:
+            # This is an unbound descriptor on the class
+            raise ValueError('cannot set unbound descriptor')
+
+        setattr(instance._parent, self.attr, value)
+
+
+class DataInfoMeta(type):
+    def __new__(mcls, name, bases, dct):
+        # Ensure that we do not gain a __dict__, which would mean
+        # arbitrary attributes could be set (as a by-product, save
+        # some memory).
+        dct.setdefault('__slots__', [])
```

Comment:
No, it has to be possible to define `__slots__` on the class (hence, the `setdefault`).

I'll remove the comment about the memory - I agree it is only confusiing.

Aside: the whole `__slots__` things is only there to disallow letting people stick random attributes on `info` (which otherwise requires overriding `__setattr__`, which makes every setting slow). Is that worth the complication?